### PR TITLE
Add kubernetes-1.29 package and variants

### PR DESCRIPTION
**Pull request by gthao313**
**Issue number:**

**Description of changes:**
This adds the package for K8s 1.29 for building kubelet and new 1.29 Bottlerocket variants.

**Testing done:**
- [x] aarch64 IPv4 conformance tests
- [x] aarch64 IPv6 conformance tests
- [x] aarch64 NVIDIA conformance tests
- [x] aarch64 NVIDIA smoke tests
- [x] x86_64 IPv4 conformance tests
- [x] x86_64 IPv6 conformance tests
- [x] x86_64 NVIDIA conformance tests
- [x] x86_64 NVIDIA smoke tests
```
[cargo-make][1] INFO - Running Task: testsys
 NAME                                                                         TYPE                                    STATE                                                                 PASSED                                FAILED                                SKIPPED   BUILD ID                                LAST UPDATE
 aarch64-aws-k8s-129-conformance-test                                         Test                                    passed                                                                   390                                     2                                   7019                                           2023-12-21T07:01:42Z
                                                                                                                                                                                                 4                                     0                                   7407
 aarch64-aws-k8s-129-ipv6-conformance-test                                    Test                                    passed                                                                   388                                     4                                   7019                                           2023-12-21T06:54:43Z
                                                                                                                                                                                                 4                                     0                                   7407
 aarch64-aws-k8s-129-nvidia-conformance-test                                  Test                                    passed                                                                   392                                     0                                   7019                                           2023-12-21T06:54:27Z
 aarch64-aws-k8s-129-nvidia-nvidia-smoke-test                                 Test                                    passed                                                                    11                                     0                                      0                                           2023-12-21T04:49:43Z
 x86-64-aws-k8s-129-conformance-test                                          Test                                    passed                                                                   392                                     0                                   7019                                           2023-12-21T00:35:26Z
 x86-64-aws-k8s-129-ipv6-conformance-test                                     Test                                    passed                                                                   392                                     0                                   7019                                           2023-12-21T00:50:31Z
 x86-64-aws-k8s-129-nvidia-conformance-test                                   Test                                    passed                                                                   392                                     0                                   7019                                           2023-12-21T01:00:53Z
 x86-64-aws-k8s-129-nvidia-nvidia-smoke-test                                  Test                                    passed                                                                    11                                     0                                      0                                           2023-12-20T23:11:53Z
 aarch64-aws-k8s-129-instances                                                Resource                                completed                                                                                                                                                                                           2023-12-21T04:40:18Z
 aarch64-aws-k8s-129-ipv6-instances                                           Resource                                completed                                                                                                                                                                                           2023-12-21T04:38:52Z
 aarch64-aws-k8s-129-nvidia-instances                                         Resource                                completed                                                                                                                                                                                           2023-12-21T04:47:11Z
 x86-64-aws-k8s-129-instances                                                 Resource                                completed                                                                                                                                                                                           2023-12-20T22:45:55Z
 x86-64-aws-k8s-129-ipv6-instances                                            Resource                                completed                                                                                                                                                                                           2023-12-20T22:52:29Z
 x86-64-aws-k8s-129-nvidia-instances                                          Resource                                completed                                                                                                                                                                                           2023-12-20T23:09:20Z
```

-  [x] metal conformance test
```
NAME                                    STATUS   ROLES           AGE     VERSION
x86-64-metal-k8s-128-blxmb              Ready    control-plane   27m     v1.29.0-eks-a5ec690
x86-64-metal-k8s-128-md-0-l6t9j-gndv8   Ready    <none>          6m      v1.29.0-eks-a5ec690
x86-64-metal-k8s-128-md-0-l6t9j-hfdkw   Ready    <none>          5m47s   v1.29.0-eks-a5ec690

➜  metal  sonobuoy results ${result}
Plugin: e2e
Status: passed
Total: 7393
Passed: 384
Failed: 0
Skipped: 7009
```

-  [x] vmware conformance test
```
➜  vmware  sonobuoy results 202312212108_sonobuoy_88b39c55-9041-4dc8-b3ba-9628e2423a55.tar.gz
Plugin: e2e
Status: passed
Total: 7393
Passed: 384
Failed: 0
Skipped: 7009
```

-  [x] EBS driver testing
```
$ kubectl --kubeconfig x86-64-aws-k8s-129-ipv6.kubeconfig  get pods -n kube-system -l app.kubernetes.io/name=aws-ebs-csi-driver
NAME                                READY   STATUS    RESTARTS   AGE
ebs-csi-controller-c5fc7f7b-2pp9h   6/6     Running   0          6m27s
ebs-csi-controller-c5fc7f7b-cbglt   6/6     Running   0          6m27s
ebs-csi-node-688vp                  3/3     Running   0          6m27s
ebs-csi-node-c7k2h                  3/3     Running   0          6m27s
```
- [x] clou-provider-credential test
Following the instruction https://github.com/bottlerocket-os/bottlerocket/pull/3329#issuecomment-1667895484
```
tianhg@ip-172-31-46-34 bottlerocket]$  kubectl get pod
NAME     READY   STATUS    RESTARTS   AGE
credential-test   1/1     Running   0          109m
``` 
**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: packages/kubernetes-1.29/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+sha512 = "fb94f967683428ca59dd8f203f398b2658ef5b65d517672d16e7f5b4e08662f9ee81065840060301e240f7659ec61543402ad4d0379c5405047c612ede8ad4eb"
+
+# RPM BuildRequires
+[build-dependencies]
+glibc = { path = "../glibc" }
+
+# RPM Requires
+[dependencies]
+aws-signing-helper = { path = "../aws-signing-helper" }
+ecr-credential-provider-1_27 = { path = "../ecr-credential-provider-1.27" }
```
Comment (etungsten) [inline packages/kubernetes-1.29/Cargo.toml]: Might want to check v1.29 of the ECR credentials provider and see if we need it or if we can stay on v1.27.

Should also consider maybe packaging the ecr credential provider together with the k8s package.

File: sources/models/src/lib.rs [hunk 2]
```diff
@@ -79,6 +79,16 @@ The `#[model]` attribute on Settings and its sub-structs reduces duplication and
 * [Model](src/aws-k8s-1.28-nvidia/mod.rs)
 * [Default settings](src/aws-k8s-1.28-nvidia/defaults.d/)
 
+## aws-k8s-1.29: Kubernetes 1.29
+
+* [Model](src/aws-k8s-1.29/mod.rs)
+* [Default settings](src/aws-k8s-1.29/defaults.d/)
```
Comment (etungsten) [inline sources/models/src/lib.rs:85]: All the links for the other variants needs to be updated as well to point to `*-k8s-1.29`. The rendered README on github does not follow symlinks.

File: packages/kubernetes-1.29/Cargo.toml (hunk 1 above)
Comment (gthao313) [reply packages/kubernetes-1.29/Cargo.toml]: I'll add v1.29 of the ECR credentials provider package and test it.

**Comment by gthao313** [comment]
New Kubernetes requires go1.21 or greater, and the go version at SDK 0.36.1 is 1.20. Therefore the CI build failed. Current We're working on bumping the Go version to 1.21. 

For the testing, I was using SDK 0.36.0 which has go1.21.

File: packages/ecr-credential-provider-1.29/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+[package]
+# "." is not allowed in crate names, but we want a friendlier name for the
+# directory and spec file, so we override it below.
+name = "ecr-credential-provider-1_27"
```
Comment (etungsten) [inline packages/ecr-credential-provider-1.29/Cargo.toml]: This is confusing. This should be 1.29 not 1.27? Same with the spec file naming.

Comment (gthao313) [reply packages/ecr-credential-provider-1.29/Cargo.toml]: yeah, it's typo. I'll fix it and re-test it.  thanks!

**Review by etungsten — APPROVED** [review]

File: sources/models/src/metal-k8s-1.29/metal-k8s-1.28 [hunk 4]
```diff
@@ -0,0 +1 @@
+metal-k8s-1.28
```
Comment (bcressey) [inline sources/models/src/metal-k8s-1.29/metal-k8s-1.28]: This is a broken symlink.

File: packages/ecr-credential-provider-1.29/Cargo.lock [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# This file is automatically @generated by Cargo.
+# It is not intended for manual editing.
+version = 3
+
+[[package]]
+name = "ecr-credential-provider-1_29"
+version = "0.1.0"
+dependencies = [
+ "glibc",
+]
```
Comment (bcressey) [inline packages/ecr-credential-provider-1.29/Cargo.lock]: This file shouldn't exist and shouldn't be committed. It's a side effect of running `cargo make` in a package subdirectory.

**Comment by gthao313** [comment]
^ addressed comments.

I'll re-test it with new SDK.

File: sources/models/README.md [hunk 6]
```diff
@@ -24,63 +24,73 @@ The `#[model]` attribute on Settings and its sub-structs reduces duplication and
 
 ### aws-k8s-1.23: Kubernetes 1.23
 
-* [Model](src/aws-k8s-1.25/mod.rs)
-* [Default settings](src/aws-k8s-1.25/defaults.d/)
+* [Model](src/aws-k8s-1.29/mod.rs)
+* [Default settings](src/aws-k8s-1.29/defaults.d/)
```
Comment (bcressey) [inline sources/models/README.md]: These aren't the correct model links for these variants.

Can you fix these to point to the same targets as the links in `sources/models/src`?

Comment (gthao313) [reply sources/models/README.md]: Yeah, Something wrong on my vscode. I'll fix it.

**Comment by gthao313** [comment]
Test result with new SKD
aws-k8s
```
 NAME                                                           TYPE                       STATE                                      PASSED                  FAILED                  SKIPPED   BUILD ID                  LAST UPDATE
 aarch64-aws-k8s-129-conformance-test                           Test                       passed                                        390                       2                     7019                             2024-01-05T21:54:58Z
                                                                                                                                           4                       0                     7407
 aarch64-aws-k8s-129-ipv6-conformance-test                      Test                       passed                                        390                       2                     7019                             2024-01-05T21:47:19Z
                                                                                                                                           4                       0                     7407
 aarch64-aws-k8s-129-nvidia-conformance-test                    Test                       passed                                        392                       0                     7019                             2024-01-06T00:39:33Z
 aarch64-aws-k8s-129-nvidia-nvidia-smoke-test                   Test                       passed                                         11                       0                        0                             2024-01-05T22:40:47Z
 x86-64-aws-k8s-129-conformance-test                            Test                       passed                                        392                       0                     7019                             2024-01-05T21:28:53Z
 x86-64-aws-k8s-129-ipv6-conformance-test                       Test                       passed                                        392                       0                     7019                             2024-01-05T21:38:36Z
 x86-64-aws-k8s-129-nvidia-conformance-test                     Test                       running                                         0                       0                        0                             2024-01-06T01:18:56Z
 x86-64-aws-k8s-129-nvidia-nvidia-smoke-test                    Test                       running                                         0                       0                        0                             2024-01-06T01:18:54Z
 aarch64-aws-k8s-129-instances                                  Resource                   completed                                                                                                                      2024-01-05T19:37:08Z
 aarch64-aws-k8s-129-ipv6-instances                             Resource                   completed                                                                                                                      2024-01-05T19:37:37Z
 aarch64-aws-k8s-129-nvidia-instances                           Resource                   completed                                                                                                                      2024-01-05T22:38:19Z
 x86-64-aws-k8s-129-instances                                   Resource                   completed                                                                                                                      2024-01-05T19:35:19Z
 x86-64-aws-k8s-129-ipv6-instances                              Resource                   completed                                                                                                                      2024-01-05T19:36:22Z
 x86-64-aws-k8s-129-nvidia-instances                            Resource                   completed                                                                                                                      2024-01-06T01:18:50Z
```

metal-k8s
```
br@br-admin:~/1.29-release/baremetal/eksa-colo$ kubectl get node --kubeconfig br-128/br-128-eks-a-cluster.kubeconfig
NAME                      STATUS   ROLES           AGE   VERSION
br-128-c94kh              Ready    control-plane   46h   v1.29.0-eks-a5ec690
br-128-md-0-7jhgq-p6zjz   Ready    <none>          46h   v1.29.0-eks-a5ec690
br-128-md-0-7jhgq-x77r2   Ready    <none>          46h   v1.29.0-eks-a5ec690

   PLUGIN     STATUS   RESULT   COUNT                                PROGRESS
      e2e   complete   passed       1   Passed:380, Failed:  0, Remaining:  0

Sonobuoy has completed. Use `sonobuoy retrieve` to get results.
Plugin: e2e
Status: passed
Total: 7393
Passed: 384
Failed: 0
Skipped: 7009
```
vmware-k8s
```
➜  vmware  sonobuoy results 202401072109_sonobuoy_fde85e40-50ff-4ea2-b99d-c9146a60bf36.tar.gz
Plugin: e2e
Status: passed
Total: 7393
Passed: 384
Failed: 0
Skipped: 7009
```

**Review by bcressey — APPROVED** [review]

File: packages/kubernetes-1.29/dockershim-symlink.conf [hunk 7]
```diff
@@ -0,0 +1,2 @@
+[Service]
```
Comment (arnaldo2792) [inline packages/kubernetes-1.29/dockershim-symlink.conf:1]: Do we still need to create this shim link?

**Review by arnaldo2792 — APPROVED** [review]

File: packages/kubernetes-1.29/dockershim-symlink.conf (hunk 7 above)
Comment (gthao313) [reply packages/kubernetes-1.29/dockershim-symlink.conf:1]: dockershim has been deprecated on 2020 but Mirantis and Docker have [committed](https://www.mirantis.com/blog/mirantis-to-take-over-support-of-kubernetes-dockershim-2/) to maintaining the dockershim after it is removed from Kubernetes. 

I think we are ok to use it, but I think it's worth to open an issue to investigate on if we should still create shim link.
https://github.com/bottlerocket-os/bottlerocket/issues/3703

**Pull request merged**